app/code/ZP/LoyaltyProgram/Controller/Adminhtml/Program/Delete.php

```diff
@@ -3,45 +3,107 @@
-use ZP\LoyaltyProgram\Api\Data\LoyaltyProgramInterface;
+use Magento\Framework\Exception\NoSuchEntityException;
 use ZP\LoyaltyProgram\Api\LoyaltyProgramRepositoryInterface;
 use Magento\Framework\Exception\LocalizedException;
-use ZP\LoyaltyProgram\Controller\Adminhtml\Program as AbstractProgramController;
+use ZP\LoyaltyProgram\Model\Controller\Adminhtml\Program\RequestHelper;
+use ZP\LoyaltyProgram\Controller\Adminhtml\Program\AbstractControllers\HttpPostActionInterface\SaveAndDelete\Controller;
+use ZP\LoyaltyProgram\Model\Controller\Adminhtml\Program\CustomerProgramManagement;
+use ZP\LoyaltyProgram\Model\LoyaltyProgram;
+use ZP\LoyaltyProgram\Model\Controller\Adminhtml\Program\SalesRuleProgramsManagement;
+use ZP\LoyaltyProgram\Model\Controller\Adminhtml\Program\Helper;
+use Psr\Log\LoggerInterface;
+use ZP\LoyaltyProgram\Api\Model\Validators\Controller\Adminhtml\Program\ValidatorInterface;
 
-class Delete extends AbstractProgramController implements HttpPostActionInterface
+class Delete extends Controller
 {
-    /**
-     * Save constructor.
-     * @param Context $context
-     * @param LoyaltyProgramRepositoryInterface $programRepository
-     */
+    public const BASIC_PROGRAM_ERR = ' DELETE ';
+
     public function __construct(
         Context $context,
-        private LoyaltyProgramRepositoryInterface $programRepository
+        LoggerInterface $logger,
+        LoyaltyProgramRepositoryInterface $programRepository,
+        CustomerProgramManagement $customerProgramManagement,
+        Helper $helper,
+        ValidatorInterface $dataValidator,
+        RequestHelper $requestHelper,
+        SalesRuleProgramsManagement $salesRuleProgramsManagement
     ) {
-        parent::__construct($context);
+        parent::__construct(
+            $context,
+            $logger,
+            $programRepository,
+            $customerProgramManagement,
+            $helper,
+            $dataValidator,
+            $requestHelper
+        );
+
+        $this->salesRuleProgramsManagement = $salesRuleProgramsManagement;
     }
 
     public function execute()
     {
-        /** @var \Magento\Backend\Model\View\Result\Redirect $resultRedirect */
-        $resultRedirect = $this->resultRedirectFactory->create();
-        $programId = $this->getRequest()->getParam(LoyaltyProgramInterface::PROGRAM_ID);
-
         try {
-            $program = $this->programRepository->get((int)$programId);
-            $programName = $program->getProgramName();
+            /** @var Redirect $resultRedirect */
+            $resultRedirect = $this->resultRedirectFactory->create();
+            $programId = $this->requestHelper->getProgramIdFromRequest($this->getRequest());
+            $this->validateProgramId($programId);
+            $this->isBasicProgram();
+            $program = $this->programRepository->get($this->programId);
+            $this->programName = $program->getProgramName();
+
+            $this->beforeDelete();
```

Comment:
No we cant, because my before and after actions which are in controllers have action logic that is not applicable to our resource model. But I will try to remember this moment and we will discuss it on our next live meeting, and may in our discussion we will find the way how to do it even more functional.